Spigot-Server-Patches/0521-Workaround-for-Client-Lag-Spikes-MC-16225.patch

```diff
@@ -0,0 +1,132 @@
+ chunk neighbours. In some specific map configurations, these calculations
+ cause a 500ms+ freeze on the Client. This patch basically serves as a
+ workaround by sending light maps to the client, so that it doesn't attempt to
+ calculate them. This mitigates the frametime impact to a minimum (but it's
+ still there).
+
+
+diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
+index 8c1f3290d23795b58a30274c9437dc7dc43fa3a1..33cca62d90bf037df1b0b4e901baedb91400aee7 100644
+--- a/src/main/java/net/minecraft/server/Chunk.java
++++ b/src/main/java/net/minecraft/server/Chunk.java
+@@ -277,7 +277,7 @@ public class Chunk implements IChunkAccess {
+ 
+                     // broadcast
+                     Object[] backingSet = inRange.getBackingSet();
+-                    Packet[] chunkPackets = new Packet[2];
++                    Packet[] chunkPackets = new Packet[10];
+                     for (int index = 0, len = backingSet.length; index < len; ++index) {
+                         Object temp = backingSet[index];
+                         if (!(temp instanceof EntityPlayer)) {
+diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+index 81bb26abc520f49de2e916cf2757ac4c710d02cd..de758f873b4736cdfb617d11fdb0d6f61e8f4a4d 100644
+--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
++++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+@@ -376,7 +376,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+                     player.needsChunkCenterUpdate = false;
+                     player.playerConnection.sendPacket(new PacketPlayOutViewCentre(currPosX, currPosZ));
+                 }
+-                PlayerChunkMap.this.sendChunk(player, new ChunkCoordIntPair(rangeX, rangeZ), new Packet[2], false, true); // unloaded, loaded
++                PlayerChunkMap.this.sendChunk(player, new ChunkCoordIntPair(rangeX, rangeZ), new Packet[10], false, true); // unloaded, loaded
+             },
+             (EntityPlayer player, int rangeX, int rangeZ, int currPosX, int currPosZ, int prevPosX, int prevPosZ,
+              com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<EntityPlayer> newState) -> {
+@@ -1955,12 +1955,92 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+ 
+     }
+ 
++    private static int getLightMask(final Chunk chunk) {
++        final ChunkSection[] chunkSections = chunk.getSections();
++        int mask = 0;
++
++        for (int i = 0; i < chunkSections.length; ++i) {
++            mask |= (ChunkSection.a(chunkSections[i]) ? 0 : 7) << i; // ChunkSection.isEmpty
++        }
++
++        return mask;
++    }
++
++    private static int getCeilingLightMask(final Chunk chunk) {
++        int mask = getLightMask(chunk);
++
++        mask |= mask >> 1;
```

Comment:
and this bithack is what? from my memory this looks like getHighestBit, but according to some documentation it appears getHighestBit is implemented differently...

Reply:
It is similar to get highest bit, it would turn an `001010` into an `001111` so basically the highest bit and all below.
We then invert this, so we'd have `110000` and compare that to the "main" chunk.
This is because the bug only appears when the current chunks lightmaps are higher than those of the neighbors, thus we can omit sending neighbors which are lower than the current chunks lights.

so TLDR is that getCeilingLightMask returns a light mask with all bits set below the highest affected section. We could also count the number of leading zeros and invert them, somehow.

Reply:
either use Integer#numberOfLeadingZeros or document what this bithack is supposed to be doing then

Reply:
I'm leaving these notes as a comment in the code, can maybe amend this in a future commit to use recommended strategy

Reply:
public static long getMask(final long value) {
        return ~(Long.MIN_VALUE >> (Long.numberOfLeadingZeros(value) - 1));
    }